### PR TITLE
RUBY-1226 Implement Drivers Sessions API and support for causally consistent reads

### PR DESCRIPTION
**Pull request by estolfo**

### CODE REVIEW

File: lib/mongo/collection.rb [hunk 1]
```diff
@@ -184,11 +190,11 @@ def create
       if (options[:collation] || options[Operation::COLLATION]) && !server.features.collation_enabled?
         raise Error::UnsupportedCollation.new
       end
-      Operation::Commands::Create.new({
-                                        selector: operation,
-                                        db_name: database.name,
-                                        write_concern: write_concern
-                                      }).execute(server)
+      Operation::Commands::Create.new(
```
Comment (saghm) [inline lib/mongo/collection.rb]: Are operations like creating and deleting collection supposed to also be done in the context of the session? I can't find anything specific in the spec about them, so I'm not sure.

File: lib/mongo/database.rb [hunk 2]
```diff
@@ -155,11 +158,11 @@ def collections
     def command(operation, opts = {})
       preference = ServerSelector.get(opts[:read] || ServerSelector::PRIMARY)
       server = preference.select_server(cluster)
-      Operation::Commands::Command.new({
+      Operation::Commands::Command.new(
```
Comment (saghm) [inline lib/mongo/database.rb]: I'm also not sure if commands should be executed in the context of a session or not

File: lib/mongo/monitoring/event/command_failed.rb [hunk 3]
```diff
@@ -55,14 +57,15 @@ class CommandFailed
         # @param [ Float ] duration The duration the command took in seconds.
         #
```
Comment (saghm) [inline lib/mongo/monitoring/event/command_failed.rb]: The `reply` parameter should probably be documented here as well.

File: lib/mongo/monitoring/event/command_failed.rb [hunk 4]
```diff
@@ -79,15 +82,16 @@ def initialize(command_name, database_name, address, request_id, operation_id, m
         # @return [ CommandFailed ] The event.
```
Comment (saghm) [inline lib/mongo/monitoring/event/command_failed.rb]: Ditto above comment about documenting `reply`

File: lib/mongo/collection.rb (hunk 1 above)
Comment (estolfo) [inline lib/mongo/collection.rb]: That's a good question - one that I asked myself as well.
The spec says 'The server reports the operation time in the response to CRUD commands.'
The ``with_session`` method extracts the operationTime from the operation response and given that the spec says the value is reported in CRUD commands, I've only wrapped CRUD commands in the ``with_session`` method. I'll keep my eye on the spec though, and test other commands, to see if operationTime is reported in more than just ones of the CRUD type.

File: lib/mongo/database.rb (hunk 2 above)
Comment (estolfo) [inline lib/mongo/database.rb]: same comment as above - I think only CRUD commands report operationTime.

File: lib/mongo/monitoring/event/command_failed.rb (hunk 3 above)
Comment (estolfo) [inline lib/mongo/monitoring/event/command_failed.rb]: thanks!

File: lib/mongo/database.rb (hunk 2 above)
Comment (xdg) [inline lib/mongo/database.rb]: It's not just CRUD.  Drivers are supposed to look for operationTime in *all* database responses, including errors.  If that's not updated in the spec, it's supposed to be.

Comment (estolfo) [inline lib/mongo/database.rb]: Thanks, I've written a comment in the google doc next to the text saying it reports it from CRUD commands.

**Pull request closed without merging**